File: go/internal/cmd/remove.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/spf13/cobra"

	"git.zib.de/cspiegel/threads/internal/git"
	"git.zib.de/cspiegel/threads/internal/thread"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	removeCommit bool
	removeMsg    string
)

var removeCmd = &cobra.Command{
	Use:               "remove <id>",
	Aliases:           []string{"rm"},
	Short:             "Remove thread entirely",
	Args:              cobra.ExactArgs(1),
	ValidArgsFunction: completeThreadIDs,
	RunE:              runRemove,
}

func init() {
	removeCmd.Flags().BoolVar(&removeCommit, "commit", false, "Commit after removing")
	removeCmd.Flags().StringVarP(&removeMsg, "m", "m", "", "Commit message")
}

func runRemove(cmd *cobra.Command, args []string) error {
	ws := getWorkspace()
	ref := args[0]

	file, err := workspace.FindByRef(ws, ref)
	if err != nil {
		return err
	}

	t, err := thread.Parse(file)
	if err != nil {
		return err
	}

	name := t.Name()
	relPath, _ := filepath.Rel(ws, file)

	// Check if file is tracked
	wasTracked := git.IsTracked(ws, relPath)

	// Remove file
	if err := os.Remove(file); err != nil {
		return fmt.Errorf("removing file: %w", err)
	}

	fmt.Printf("Removed: %s\n", file)

	if !wasTracked {
		fmt.Println("Note: Thread was never committed to git, no commit needed.")
		return nil
	}

	if removeCommit {
		msg := removeMsg
		if msg == "" {
			msg = fmt.Sprintf("threads: remove '%s'", name)
		}
		if err := git.Add(ws, relPath); err != nil {
			return err
		}
		if err := git.Commit(ws, []string{relPath}, msg); err != nil {
			return err
		}
		if err := git.Push(ws); err != nil {
			fmt.Printf("WARNING: git push failed (commit succeeded): %v\n", err)
		}
	} else {
		fmt.Println("Note: To commit this removal, run:")
		fmt.Printf("  git -C \"$WORKSPACE\" add \"%s\" && git -C \"$WORKSPACE\" commit -m \"threads: remove '%s'\"\n", relPath, name)
	}

	return nil
}

File: rust/src/cmd/remove.rs
use std::fs;
use std::path::Path;

use clap::Args;
use clap_complete::engine::ArgValueCompleter;

use crate::git;
use crate::thread::Thread;
use crate::workspace;

#[derive(Args)]
pub struct RemoveArgs {
    /// Thread ID or name reference
    #[arg(add = ArgValueCompleter::new(crate::workspace::complete_thread_ids))]
    id: String,

    /// Commit after removing
    #[arg(long)]
    commit: bool,

    /// Commit message
    #[arg(short = 'm', long)]
    m: Option<String>,
}

pub fn run(args: RemoveArgs, ws: &Path) -> Result<(), String> {
    let file = workspace::find_by_ref(ws, &args.id)?;

    let t = Thread::parse(&file)?;
    let name = t.name().to_string();
    let rel_path = file
        .strip_prefix(ws)
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_else(|_| file.to_string_lossy().to_string());

    // Check if file is tracked
    let was_tracked = git::is_tracked(ws, &rel_path);

    // Remove file
    fs::remove_file(&file)
        .map_err(|e| format!("removing file: {}", e))?;

    println!("Removed: {}", file.display());

    if !was_tracked {
        println!("Note: Thread was never committed to git, no commit needed.");
        return Ok(());
    }

    if args.commit {
        let msg = args.m.unwrap_or_else(|| format!("threads: remove '{}'", name));
        git::add(ws, &[&rel_path])?;
        git::commit(ws, &[rel_path.clone()], &msg)?;
        eprintln!("Note: Changes are local. Push with 'git push' when ready.");
    } else {
        println!("Note: To commit this removal, run:");
        println!(
            "  git -C \"$WORKSPACE\" add \"{}\" && git -C \"$WORKSPACE\" commit -m \"threads: remove '{}'\"",
            rel_path, name
        );
    }

    Ok(())
}
